### PR TITLE
Set a numbered variable for each acquired lock

### PR DESCRIPTION
**Pull request by gaspardpetit**
This is a step towards implementing the feature from https://github.com/jenkinsci/lockable-resources-plugin/pull/20 where lock properties can be configured and exposed to the pipeline.

One of the issue with this new feature is that in some cases, multiple locks are acquired, making the exposition of properties cumbersome.

This alters the behaviour of the `variable` extraction in the following way:

- The environment variable with the name of the requested `variable` will still contain the lock name, or a comma separated list of lock names; *this does not break compatibility*
- Additional environment variables are set, with the `variable` name as a prefix and a number as a suffix, making it possible to retrieve individual lock names, ex:

```groovy
lock(label: 'some_resource', variable: 'LOCKED_RESOURCE', quantity: 2) {
  // comma separated names of all acquired locks
  echo env.LOCKED_RESOURCE

  // first lock
  echo env.LOCKED_RESOURCE0

  // second lock
  echo env.LOCKED_RESOURCE1
}
```

If this is approved, I will submit another PR to expose lock properties with the `<name><number>_<propertykey>` format.

This differs from the approach in https://github.com/jenkinsci/lockable-resources-plugin/pull/20 where new arguments were added to the plugin: `injectProperties`, `prefixProperties` and `uppercaseProperties`

In my implementation, instead I will simplify this to assume that when `variable` is set, the properties are also requested and should use the `variable` value (with index) for properties, ex.

```groovy
lock(label: 'some_resource', variable: 'LOCKED_RESOURCE') {
 // name of the lock
  echo env.LOCKED_RESOURCE0
 // value of the property KEY_A for the acquired lock
  echo env.LOCKED_RESOURCE0_KEY_A
}
```

I don't think it is necessary to make the feature more complex than this to make it useful.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java
Comment (jimklimov) [inline src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java:838]: This PR description mentions lack of `uppercaseProperties` compared to original PR it replays.
So just to clarify: here the variable names (prefix part at least) keeps the casing of the `lock(variable:...)` argument. Whatever the pipeline script author spelled, right?

**Review by jimklimov — APPROVED** [review]

**Comment by jimklimov** [comment]
After #293, can you please revise the merge-conflicts?

**Comment by jimklimov** [comment]
@gaspardpetit : hello, while looking at issue history, I saw a number of questions that imply that the order of resource names in the environment variable(s) does not always match the order of `extra` map. I understand this differs from the problem you were solving here with locking some quantity of same resources by label, but would you please be able to look into that problem too?

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java
Comment (gaspardpetit) [reply src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java:838]: Correct, while it's a good practice to keep environment variable in uppercase, I think I'd leave it up to whoever decides to name their property to follow that or not. If we enforce upper-casing (or provide an option for it), I would argue that this should apply to the current `variable` or it starts getting weird. For example, what would be expected of

```
lock(label: 'label1', variable: 'someVar', `uppercaseProperties: true) {
  // should we use env.someVar or env.SOMEVAR ?
}
```

There are also many situations where the properties could end up as invalid environment variable names, such as using UTF8 characters, spaces, hyphens, etc. But since this wasn't handled for `variable` I chose to not handle it and leave it to the users to understand that if they use a emoji in the name of their variable it just... won't work

**Comment by gaspardpetit** [comment]
I'll take a look at the extra-map ordering, it's indeed in the same area as what I've been working on, so shouldn't be to difficult to fix.

**Comment by gaspardpetit** [comment]
Thank you for quick actions on this PR, I really appreciate!

**Comment by basil** [comment]
Released in [2.15](https://github.com/jenkinsci/lockable-resources-plugin/releases/tag/lockable-resources-2.15).

**Pull request merged**
